### PR TITLE
CLOUDP-193669: Make overriden tolerations take precedence

### PR DESCRIPTION
**Pull request by mircea-cosbuc**
This patch addresses a bug where overriding tolerations in the StatefulSet spec would end up in an endless reconciliation loop. 

Addresses:  
- https://github.com/mongodb/mongodb-kubernetes-operator/issues/1344
- https://github.com/mongodb/mongodb-kubernetes-operator/issues/1349

### CODE REVIEW

**Review by slaskawi — APPROVED** [review]
LGTM!

One thing you might think of is whether this fix deserves an e2e test. To me personally, a unit test would be sufficient. But I'm leaving this up to you - great work!

**Comment by mircea-cosbuc** [comment]
@slaskawi The STS override e2e tests were updated to include tolerations.

File: pkg/kube/podtemplatespec/podspec_template_test.go [hunk 1]
```diff
@@ -247,6 +247,140 @@ func TestMergeEnvironmentVariables(t *testing.T) {
...
+					Operator: corev1.TolerationOpEqual,
+				},
+			},
+			overrideTolerations: []corev1.Toleration{},
+			expectedTolerations: []corev1.Toleration{},
+		},
+		{
+			// Overriding toleration should replace a nil original toleration.
+			name:               "Overriding when default toleration is nil",
+			defaultTolerations: nil,
```
Comment (lucian-tosa) [inline pkg/kube/podtemplatespec/podspec_template_test.go:304]: Would it make sense to also test when defaultTolerations is an empty array?

File: test/e2e/statefulset_arbitrary_config/statefulset_arbitrary_config_test.go [hunk 2]
```diff
@@ -50,4 +67,20 @@ func TestStatefulSetArbitraryConfig(t *testing.T) {
 	t.Run("Container has been merged by name", mongodbtests.StatefulSetContainerConditionIsTrue(&mdb, "mongodb-agent", func(container corev1.Container) bool {
 		return container.ReadinessProbe.TimeoutSeconds == 100
 	}))
+	t.Run("Tolerations have been added correctly", mongodbtests.StatefulSetConditionIsTrue(&mdb, func(sts appsv1.StatefulSet) bool {
```
Comment (lucian-tosa) [inline test/e2e/statefulset_arbitrary_config/statefulset_arbitrary_config_test.go:70]: It seems that this test doesn't check whether tolerations are added only once. Since this was the initial issue, maybe this deserves an e2e. Correct me if I'm wrong.

File: pkg/kube/podtemplatespec/podspec_template_test.go [hunk 3]
```diff
@@ -247,6 +247,140 @@ func TestMergeEnvironmentVariables(t *testing.T) {
...
+				},
+				{
+					Key:      "key1",
+					Value:    "value2",
+					Operator: corev1.TolerationOpExists,
+				},
+			},
+		},
+		{
+			// Overriding toleration should replace any original toleration.
```
Comment (lucian-tosa) [inline pkg/kube/podtemplatespec/podspec_template_test.go:331]: Is this the correct behaviour? As I noticed, the behaviour before the bug was introduced was more of "merge" instead of a total override.

Comment (mircea-cosbuc) [reply pkg/kube/podtemplatespec/podspec_template_test.go:331]: Yes, this is the direction of this PR. Merging the tolerations only makes sense if the operator sets tolerations when creating the statefulsets, which it doesn't since it has no bias about the cluster topology. We want whatever is in the override to be reflected in the statefulset.

File: test/e2e/statefulset_arbitrary_config/statefulset_arbitrary_config_test.go (hunk 2 above)
Comment (mircea-cosbuc) [reply test/e2e/statefulset_arbitrary_config/statefulset_arbitrary_config_test.go:70]: Good point, this should actually check if the tolerations are identical since only the pods get the default kubernetes tolerations added to them so it's not necessary to iterate and check if they are contained.

File: pkg/kube/podtemplatespec/podspec_template_test.go (hunk 1 above)
Comment (mircea-cosbuc) [reply pkg/kube/podtemplatespec/podspec_template_test.go:304]: This is identical to the case when `defaultTolerations` is a non-empty array, since we discard everything there. Should we change the logic, that test might make sense.

**Comment by lucian-tosa** [comment]
LGTM

**Review by lucian-tosa — APPROVED** [review]
